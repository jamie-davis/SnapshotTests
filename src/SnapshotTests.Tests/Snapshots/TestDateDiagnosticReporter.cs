using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using NUnit.Framework;
using SnapshotTests.Snapshots;
using TestConsoleLib;
using TestConsoleLib.Testing;

namespace SnapshotTests.Tests.Snapshots
{
    [TestFixture]
    public class TestDateDiagnosticReporter
    {
        [Test]
        public void DateDiagnosticsAreReported()
        {
            //Arrange
            var collection = new SnapshotCollection();

            var beforeDates = new[]
            {
                DateTime.Parse("2020-10-11 10:35"),
                DateTime.Parse("2020-10-11 10:36"),
                DateTime.Parse("2020-10-11 10:37"),
                DateTime.Parse("2020-10-11 10:38"),
            };

            var afterDates = new[]
            {
                DateTime.Parse("2020-10-11 10:35"),
                DateTime.Parse("2020-10-11 10:36"),
                DateTime.Parse("2020-10-11 10:39"),
                DateTime.Parse("2020-10-11 10:40"),
                DateTime.Parse("2020-10-11 10:41"),
            };

            var beforeBuilder = collection.NewSnapshot("before");

            collection.DefineTable("Dates").PrimaryKey("Key").IsUnpredictable("Date");

            beforeDates.Select((bd, ix) => new { Key = ix, Date = bd, Other = "o", OtherDate = DateTime.MinValue}).ToSnapshotTable(beforeBuilder, "Dates");
            var deleteRow = beforeBuilder.AddNewRow("Dates");
            deleteRow["Key"] = 100;
            deleteRow["Date"] = DateTime.Parse("2020-10-18 11:19");
            deleteRow["Other"] = "o";
            deleteRow["OtherDate"] = DateTime.MinValue;
            
            var afterBuilder = collection.NewSnapshot("after");
            
            afterDates.Select((bd, ix) => new { Key = ix, Date = bd, Other = "a", OtherDate = DateTime.MaxValue}).ToSnapshotTable(afterBuilder, "Dates");

            var before = collection.GetSnapshot("before");
            var after = collection.GetSnapshot("after");
            var differences = SnapshotDifferenceAnalyser.ExtractDifferences(collection, before, after);

            var output = new Output();
            
            //Act
            var ranges = new List<NamedTimeRange>
            {
                new NamedTimeRange(beforeDates[0], beforeDates[3], "before"),
                new NamedTimeRange(afterDates[2], afterDates[3], "after"),
            };
            DateDiagnosticReporter.Report(output, ranges, before, after, differences);

            //Assert
            output.Report.Verify();
        }
    }
}
