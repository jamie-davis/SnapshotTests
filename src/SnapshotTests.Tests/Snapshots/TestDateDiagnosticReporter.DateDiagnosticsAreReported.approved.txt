====> Date diagnostics begin                                                                   
Snapshot date ranges                                                                           
--------------------                                                                           
Snapshot Start               End                                                               
-------- ------------------- -------------------                                               
before   11/10/2020 10:35:00 11/10/2020 10:38:00                                               
after    11/10/2020 10:39:00 11/10/2020 10:40:00                                               
                                                                                               
Dates From Difference Set                                                                      
-------------------------                                                                      
Table                                                                                          
Name                                                                                           
-----                                                                                          
Dates                                                                                          
                                                                                               
    Key OtherDate Before    OtherDate After     Date Before                  Date After                  
    --- ------------------- ------------------- ---------------------------- ----------------------------
    0   01/01/0001 00:00:00 31/12/9999 23:59:59 Date-1 (11/10/2020 10:35:00) Date-1 (11/10/2020 10:35:00)
    1   01/01/0001 00:00:00 31/12/9999 23:59:59 Date-2 (11/10/2020 10:36:00) Date-2 (11/10/2020 10:36:00)
    2   01/01/0001 00:00:00 31/12/9999 23:59:59 Date-3 (11/10/2020 10:37:00) Date-4 (11/10/2020 10:39:00)
    3   01/01/0001 00:00:00 31/12/9999 23:59:59 Date-5 (11/10/2020 10:38:00) Date-6 (11/10/2020 10:40:00)
    4                       31/12/9999 23:59:59                              Date-7 (11/10/2020 10:41:00)
    100 01/01/0001 00:00:00                     Date-8 (18/10/2020 11:19:00)                             
                                                                                               
====> Date diagnostics end                                                                     
                                                                                               